File: src/lib/fatigueIndex.ts
/**
 * fatigueIndex.ts
 *
 * Calculates accumulated weekly training volume per muscle group and compares
 * it to research-based Maximum Recoverable Volume (MRV) values.
 *
 * Reference ranges are conservative medians from Israetel et al. (RP Strength).
 * The actual MRV varies by individual training age, recovery, nutrition, etc.
 * These values serve as evidence-based starting points.
 */

import type { WorkoutLog, WorkoutDay } from '../store/useAppStore';
import { format, startOfWeek, endOfWeek } from 'date-fns';

/**
 * MRV = Maximum Recoverable Volume (sets per week per muscle group).
 * Below this threshold the athlete recovers fully between sessions.
 * Values are weekly SET counts per muscle group.
 */
export const MRV_DEFAULTS: Record<string, number> = {
  Pecho: 20,
  Espalda: 25,
  Hombros: 20,
  Piernas: 22,
  Bíceps: 20,
  Tríceps: 20,
  Core: 25,
  Calistenia: 18,
  Glúteos: 20,
  Cardio: 5,
};

const DEFAULT_MRV = 18;

export type FatigueEntry = {
  muscleGroup: string;
  weeklyVolume: number;  // total sets this week
  mrv: number;           // max recoverable volume (sets)
  percent: number;       // 0-100+ percent of MRV used
  status: 'fresh' | 'moderate' | 'high' | 'overreached';
};

/**
 * Returns fatigue status per muscle group for the current ISO week (Monday-based).
 *
 * @param logs       All stored workout logs
 * @param routine    Weekly routine (to map exerciseId → muscleGroup)
 * @param weekStart  Reference date for the start of the target week (defaults to today's week)
 */
export function computeFatigueIndex(
  logs: WorkoutLog[],
  routine: WorkoutDay[],
  weekStart?: Date,
): FatigueEntry[] {
  const now = weekStart ?? new Date();
  const monday = startOfWeek(now, { weekStartsOn: 1 });
  const sunday = endOfWeek(now, { weekStartsOn: 1 });
  const fromKey = format(monday, 'yyyy-MM-dd');
  const toKey = format(sunday, 'yyyy-MM-dd');

  // Build exerciseId → muscleGroup lookup from routine
  const muscleMap = new Map<string, string>();
  for (const day of routine) {
    for (const ex of day.exercises) {
      if (ex.id && ex.muscleGroup) {
        muscleMap.set(ex.id, ex.muscleGroup);
      }
    }
  }

  // Count sets per muscle group this week
  const setsByMuscle = new Map<string, number>();
  for (const log of logs) {
    const dateKey = log.date.slice(0, 10);
    if (dateKey < fromKey || dateKey > toKey) continue;

    const muscle = muscleMap.get(log.exerciseId);
    if (!muscle) continue;

    setsByMuscle.set(muscle, (setsByMuscle.get(muscle) ?? 0) + 1);
  }

  const result: FatigueEntry[] = [];

  for (const [muscle, sets] of setsByMuscle.entries()) {
    const mrv = MRV_DEFAULTS[muscle] ?? DEFAULT_MRV;
    const percent = Math.round((sets / mrv) * 100);
    let status: FatigueEntry['status'];

    if (percent < 50) status = 'fresh';
    else if (percent < 75) status = 'moderate';
    else if (percent < 100) status = 'high';
    else status = 'overreached';

    result.push({ muscleGroup: muscle, weeklyVolume: sets, mrv, percent, status });
  }

  // Sort by percent descending (most fatigued first)
  return result.sort((a, b) => b.percent - a.percent);
}

/** Human-readable status label */
export function fatigueStatusLabel(status: FatigueEntry['status']): string {
  switch (status) {
    case 'fresh': return '🟢 Fresco';
    case 'moderate': return '🟡 Moderado';
    case 'high': return '🟠 Alto';
    case 'overreached': return '🔴 Sobreentrenado';
  }
}

/** CSS colour class for the status badge */
export function fatigueStatusColor(status: FatigueEntry['status']): string {
  switch (status) {
    case 'fresh': return 'text-emerald-400';
    case 'moderate': return 'text-yellow-400';
    case 'high': return 'text-orange-400';
    case 'overreached': return 'text-red-400';
  }
}

File: voltbody-android/app/src/main/java/com/voltbody/app/domain/usecase/FatigueIndex.kt
package com.voltbody.app.domain.usecase

import com.voltbody.app.domain.model.WorkoutDay
import com.voltbody.app.domain.model.WorkoutLog
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import java.time.temporal.WeekFields
import java.util.Locale

// ── Fatigue Index (ported from fatigueIndex.ts) ───────────────────────────────

val MRV_DEFAULTS = mapOf(
    "Pecho" to 20,
    "Espalda" to 25,
    "Hombros" to 20,
    "Piernas" to 22,
    "Bíceps" to 20,
    "Tríceps" to 20,
    "Core" to 25,
    "Calistenia" to 18,
    "Glúteos" to 20,
    "Cardio" to 5
)
private const val DEFAULT_MRV = 18

enum class FatigueStatus { FRESH, MODERATE, HIGH, OVERREACHED }

data class FatigueEntry(
    val muscleGroup: String,
    val weeklyVolume: Int,
    val mrv: Int,
    val percent: Int,
    val status: FatigueStatus
)

fun fatigueStatusLabel(status: FatigueStatus) = when (status) {
    FatigueStatus.FRESH -> "🟢 Fresco"
    FatigueStatus.MODERATE -> "🟡 Moderado"
    FatigueStatus.HIGH -> "🟠 Alto"
    FatigueStatus.OVERREACHED -> "🔴 Sobreentrenado"
}

fun computeFatigueIndex(
    logs: List<WorkoutLog>,
    routine: List<WorkoutDay>,
    weekStart: LocalDate = LocalDate.now()
): List<FatigueEntry> {
    val wf = WeekFields.of(Locale("es"))
    val monday = weekStart.with(wf.dayOfWeek(), 1)
    val sunday = monday.plusDays(6)
    val fromKey = monday.format(DateTimeFormatter.ISO_LOCAL_DATE)
    val toKey = sunday.format(DateTimeFormatter.ISO_LOCAL_DATE)

    val muscleMap = buildMap {
        for (day in routine) {
            for (ex in day.exercises) {
                if (ex.id.isNotEmpty() && ex.muscleGroup.isNotEmpty()) put(ex.id, ex.muscleGroup)
            }
        }
    }

    val setsByMuscle = mutableMapOf<String, Int>()
    for (log in logs) {
        val dateKey = log.date.take(10)
        if (dateKey < fromKey || dateKey > toKey) continue
        val muscle = muscleMap[log.exerciseId] ?: continue
        setsByMuscle[muscle] = (setsByMuscle[muscle] ?: 0) + 1
    }

    return setsByMuscle.map { (muscle, sets) ->
        val mrv = MRV_DEFAULTS[muscle] ?: DEFAULT_MRV
        val percent = ((sets.toFloat() / mrv) * 100).toInt()
        val status = when {
            percent < 50 -> FatigueStatus.FRESH
            percent < 75 -> FatigueStatus.MODERATE
            percent < 100 -> FatigueStatus.HIGH
            else -> FatigueStatus.OVERREACHED
        }
        FatigueEntry(muscle, sets, mrv, percent, status)
    }.sortedByDescending { it.percent }
}
